### ISSUE TITLE
Compile project for Android

### ISSUE THREAD

**Issue by andreas-eberle** [deeplearning4j/libnd4j#208]
Please make libnd4j available for Android.

**Comment by andreas-eberle** [comment]
I was able to compile libnd4j for armeabi-v7a (I just got it to work... it's hacky...). However, after I build nd4j and included it with the so files into my application, I get the following error:

```
05-27 23:51:31.582 31476-31993/? W/linker: /data/app/de.demo.app-1/lib/arm/libnd4j.so: unused DT entry: type 0xf arg 0x26c28
05-27 23:51:31.582 31476-31993/? E/AndroidRuntime: FATAL EXCEPTION: Thread-5695
                                                   Process: de.demo.app, PID: 31476
                                                   java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/de.demo.app-1/base.apk"],nativeLibraryDirectories=[/data/app/de.demo.app-1/lib/arm, /data/app/de.demo.app-1/base.apk!/lib/armeabi-v7a, /vendor/lib, /system/lib]]] couldn't find "libjnind4j.so"
                                                       at java.lang.Runtime.loadLibrary(Runtime.java:367)
                                                       at java.lang.System.loadLibrary(System.java:1076)
                                                       at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:702)
                                                       at org.bytedeco.javacpp.Loader.load(Loader.java:500)
                                                       at org.nd4j.nativeblas.NativeOps.<clinit>(NativeOps.java:30)
                                                       at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.<init>(NativeOpExecutioner.java:29)
                                                       at java.lang.Class.newInstance(Native Method)
                                                       at org.nd4j.linalg.factory.Nd4j.initWithBackend(Nd4j.java:4773)
                                                       at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:4722)
                                                       at org.nd4j.linalg.factory.Nd4j.<clinit>(Nd4j.java:145)
                                                       at org.deeplearning4j.nn.params.ConvolutionParamInitializer.createBias(ConvolutionParamInitializer.java:65)
                                                       at org.deeplearning4j.nn.params.ConvolutionParamInitializer.init(ConvolutionParamInitializer.java:48)
                                                       at org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.getParams(DefaultLayerFactory.java:107)
                                                       at org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.create(DefaultLayerFactory.java:64)
                                                       at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.init(MultiLayerNetwork.java:361)
                                                    Caused by: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/de.demo.app-1/base.apk"],nativeLibraryDirectories=[/data/app/de.demo.app-1/lib/arm, /data/app/de.demo.app-1/base.apk!/lib/armeabi-v7a, /vendor/lib, /system/lib]]] couldn't find "libgnustl_static.so"
                                                       at java.lang.Runtime.loadLibrary(Runtime.java:367)
                                                       at java.lang.System.loadLibrary(System.java:1076)
                                                       at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:702)
                                                       at org.bytedeco.javacpp.Loader.load(Loader.java:491)
                                                       at org.nd4j.nativeblas.NativeOps.<clinit>(NativeOps.java:30) 
                                                       at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.<init>(NativeOpExecutioner.java:29) 
                                                       at java.lang.Class.newInstance(Native Method) 
                                                       at org.nd4j.linalg.factory.Nd4j.initWithBackend(Nd4j.java:4773) 
                                                       at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:4722) 
                                                       at org.nd4j.linalg.factory.Nd4j.<clinit>(Nd4j.java:145) 
                                                       at org.deeplearning4j.nn.params.ConvolutionParamInitializer.createBias(ConvolutionParamInitializer.java:65) 
                                                       at org.deeplearning4j.nn.params.ConvolutionParamInitializer.init(ConvolutionParamInitializer.java:48) 
                                                       at org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.getParams(DefaultLayerFactory.java:107) 
                                                       at org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.create(DefaultLayerFactory.java:64) 
                                                       at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.init(MultiLayerNetwork.java:361) 
```

For me, it looks like the libnd4j.so has been found and loaded. However, it seems to require another lib called libgnustl_static.so which it cannot find.

@saudet: I googled the error message and found some old bugs and threads suggesting to change the proguard settings (https://github.com/bytedeco/javacv/wiki/Configuring-Proguard-for-JavaCV). However, that didn't help either.

Do you have an idea what I'm doing wrong?

**Comment by saudet** [comment]
No, it doesn't require "libgnustl_static.so". That's a bogus guess. It usually means that Android wasn't able to load the library because of missing symbols. That shows up in the Android log usually. Unlike normal Linux, if there is even one symbol it can't find, it won't load it. We can detect if there are any missing symbols at link time by using the `-Wl,--no-undefined` option at link time.

**Comment by andreas-eberle** [comment]
@saudet: I tried the flags you suggested, but it didn't show errors/warnings during linking. However, I managed to get rid of that exception by also adding the libjnind4j.so generated by the nd4j project.

But now I get the following exception:
```
05-30 09:56:35.806 13502-14289/de.demo E/AndroidRuntime: FATAL EXCEPTION: Thread-1920
  Process: de.demo, PID: 13502
  java.lang.UnsatisfiedLinkError: dlopen failed: could not load library "libnd4j.so" needed by "libjnind4j.so"; caused by cannot locate symbol "_Znwj" referenced by "libnd4j.so"...
	  at java.lang.Runtime.loadLibrary(Runtime.java:371)
	  at java.lang.System.loadLibrary(System.java:989)
	  at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:702)
	  at org.bytedeco.javacpp.Loader.load(Loader.java:500)
	  at org.nd4j.nativeblas.NativeOps.<clinit>(NativeOps.java:30)
	  at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.<init>(NativeOpExecutioner.java:29)
	  at java.lang.reflect.Constructor.newInstance(Native Method)
	  at java.lang.Class.newInstance(Class.java:1650)
	  at org.nd4j.linalg.factory.Nd4j.initWithBackend(Nd4j.java:4773)
	  at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:4722)
	  at org.nd4j.linalg.factory.Nd4j.<clinit>(Nd4j.java:145)
	  at org.deeplearning4j.nn.params.ConvolutionParamInitializer.createBias(ConvolutionParamInitializer.java:65)
	  at org.deeplearning4j.nn.params.ConvolutionParamInitializer.init(ConvolutionParamInitializer.java:48)
	  at org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.getParams(DefaultLayerFactory.java:107)
	  at org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.create(DefaultLayerFactory.java:64)
	  at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.init(MultiLayerNetwork.java:361)
   Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_Znwj" referenced by "libnd4j.so"...
	  at java.lang.Runtime.loadLibrary(Runtime.java:371)
	  at java.lang.System.loadLibrary(System.java:989)
	  at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:702)
	  at org.bytedeco.javacpp.Loader.load(Loader.java:491)
	  at org.nd4j.nativeblas.NativeOps.<clinit>(NativeOps.java:30) 
	  at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.<init>(NativeOpExecutioner.java:29) 
	  at java.lang.reflect.Constructor.newInstance(Native Method) 
	  at java.lang.Class.newInstance(Class.java:1650) 
	  at org.nd4j.linalg.factory.Nd4j.initWithBackend(Nd4j.java:4773) 
	  at org.nd4j.linalg.factory.Nd4j.initContext(Nd4j.java:4722) 
	  at org.nd4j.linalg.factory.Nd4j.<clinit>(Nd4j.java:145) 
	  at org.deeplearning4j.nn.params.ConvolutionParamInitializer.createBias(ConvolutionParamInitializer.java:65) 
	  at org.deeplearning4j.nn.params.ConvolutionParamInitializer.init(ConvolutionParamInitializer.java:48) 
	  at org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.getParams(DefaultLayerFactory.java:107) 
	  at org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.create(DefaultLayerFactory.java:64) 
	  at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.init(MultiLayerNetwork.java:361) 
```

Do you have an idea why libnd4j is missing a symbol? Or how I can find out which symbol that really is? Do you think this is a symbol that should be available in libnd4j or is it something in another library that should be there?

Many thanks in advance!

**Comment by andreas-eberle** [comment]
@saudet: I managed t get errors with the flag you posted above:

```
$ /D/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/arm-linux-androideabi-gcc.exe  -fPIC -fexceptions -frtti -fpic -Wno-psabi --sysroot=D:/programs/Android/SDK/ndk-bundle/platforms/andro
id-14/arch-arm -funwind-tables -finline-limit=64 -fsigned-char -no-canonical-prefixes -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -fdata-sections -ffunction-sections -Wa,--noexecstack  -marm -mfpu=neon -mfloat-abi=hard -march=armv
7-a -Wl,--no-warn-mismatch -Wall -g -Wl,-rpath,RIGIN/ -fopenmp -Wall -O3 -std=c++11 -fassociative-math -funsafe-math-optimizations -lm_hard -shared -Wl,-soname,libnd4j.so -o libnd4j.so "CMakeFiles/nd4j.dir/cpu/NativeBlas.cpp.o" "CMakeFi
les/nd4j.dir/cpu/NativeOps.cpp.o"   -L/libs/armeabi-v7a  -L/d/test/Beispielproject/libnd4j/OpenBlas/lib /d/test/Beispielproject/libnd4j/OpenBlas/lib/libopenblas.a -symbolic -lm -Wl,--no-undefined
D:/test/Beispielproject/libnd4j/include/reduce.h:1737: error: undefined reference to 'operator new(unsigned int)'
D:/test/Beispielproject/libnd4j/include/reduce.h:1683: error: undefined reference to 'operator new(unsigned int)'
D:/test/Beispielproject/libnd4j/include/reduce.h:1701: error: undefined reference to 'operator new(unsigned int)'
D:/test/Beispielproject/libnd4j/include/reduce.h:1689: error: undefined reference to 'operator new(unsigned int)'
D:/test/Beispielproject/libnd4j/include/reduce.h:1748: error: undefined reference to 'operator delete(void*)'
D:/test/Beispielproject/libnd4j/include/reduce.h:1748: error: undefined reference to 'operator delete(void*)'
D:/test/Beispielproject/libnd4j/include/reduce.h:1085: error: undefined reference to 'operator delete(void*)'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:function virtual thunk to functions::reduce::ops::Mean<double>::~Mean(): error: undefined reference to 'operator delete(void*)'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o(.ARM.extab.text._ZN9functions6reduce14ReduceFunctionIdE4execEPdPiS3_S3_S4_S4_i._omp_fn.0+0x0): error: undefined reference to '__gxx_personality_v0'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o(.ARM.extab.text._ZN9functions6reduce14ReduceFunctionIdE4execEPdPiS3_S3_S4_S4_i._omp_fn.1+0x0): error: undefined reference to '__gxx_personality_v0'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o(.ARM.extab.text._ZN9functions19pairwise_transforms17PairWiseTransformIdE4execEPdxS3_xS3_xS3_x+0x0): error: undefined reference to '__gxx_personality_v0'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o(.ARM.extab.text._ZN9functions9transform9TransformIdE4execEPdiS3_iS3_i+0x0): error: undefined reference to '__gxx_personality_v0'
D:/test/Beispielproject/libnd4j/include/reduce3.h:911: error: undefined reference to 'operator new[](unsigned int)'
D:/test/Beispielproject/libnd4j/include/reduce3.h:911: error: undefined reference to 'operator new[](unsigned int)'
D:/test/Beispielproject/libnd4j/include/reduce3.h:1307: error: undefined reference to 'operator delete[](void*)'
D:/test/Beispielproject/libnd4j/include/reduce3.h:1182: error: undefined reference to 'operator delete[](void*)'
D:/test/Beispielproject/libnd4j/include/reduce3.h:919: error: undefined reference to 'operator delete[](void*)'
D:/test/Beispielproject/libnd4j/include/reduce3.h:1057: error: undefined reference to 'operator delete[](void*)'
D:/test/Beispielproject/libnd4j/include/reduce3.h:859: error: undefined reference to 'operator new[](unsigned int)'
D:/test/Beispielproject/libnd4j/include/reduce3.h:859: error: undefined reference to 'operator new[](unsigned int)'
D:/test/Beispielproject/libnd4j/include/indexreduce.h:755: error: undefined reference to '__cxa_end_cleanup'
D:/test/Beispielproject/libnd4j/include/summarystatsreduce.h:1032: error: undefined reference to '__cxa_end_cleanup'
D:/test/Beispielproject/libnd4j/include/indexreduce.h:755: error: undefined reference to '__cxa_end_cleanup'
D:/test/Beispielproject/libnd4j/include/summarystatsreduce.h:1032: error: undefined reference to '__cxa_end_cleanup'
D:/programs/Android/SDK/ndk-bundle/sources/cxx-stl/gnu-libstdc++/4.9/include/ext/new_allocator.h:102: error: undefined reference to 'std::__throw_bad_alloc()'
D:/programs/Android/SDK/ndk-bundle/sources/cxx-stl/gnu-libstdc++/4.9/include/ext/new_allocator.h:102: error: undefined reference to 'std::__throw_bad_alloc()'
D:/programs/Android/SDK/ndk-bundle/sources/cxx-stl/gnu-libstdc++/4.9/include/ext/new_allocator.h:102: error: undefined reference to 'std::__throw_bad_alloc()'
D:/programs/Android/SDK/ndk-bundle/sources/cxx-stl/gnu-libstdc++/4.9/include/ext/new_allocator.h:102: error: undefined reference to 'std::__throw_bad_alloc()'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::ops::OpFactory<double>: error: undefined reference to 'vtable for __cxxabiv1::__class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::ops::Op<double>: error: undefined reference to 'vtable for __cxxabiv1::__class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::reduce::ReduceFunction<double>: error: undefined reference to 'vtable for __cxxabiv1::__si_class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::reduce::ReduceOpFactory<double>: error: undefined reference to 'vtable for __cxxabiv1::__vmi_class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::ops::OpFactory<float>: error: undefined reference to 'vtable for __cxxabiv1::__class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::ops::Op<float>: error: undefined reference to 'vtable for __cxxabiv1::__class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::reduce::ReduceFunction<float>: error: undefined reference to 'vtable for __cxxabiv1::__si_class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::reduce::ReduceOpFactory<float>: error: undefined reference to 'vtable for __cxxabiv1::__vmi_class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::indexreduce::IndexReduce<double>: error: undefined reference to 'vtable for __cxxabiv1::__si_class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::broadcast::Broadcast<double>: error: undefined reference to 'vtable for __cxxabiv1::__si_class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::pairwise_transforms::PairWiseTransform<double>: error: undefined reference to 'vtable for __cxxabiv1::__vmi_class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:typeinfo for functions::reduce3::Reduce3<double>: error: undefined reference to 'vtable for __cxxabiv1::__vmi_class_type_info'
D:/programs/Android/SDK/ndk-bundle/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.exe: the vtable symbol may be undefined because the class is missing its key function (see go/missingkeymethod)
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:vtable for functions::ops::OpFactory<double>: error: undefined reference to '__cxa_pure_virtual'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:vtable for functions::ops::Op<double>: error: undefined reference to '__cxa_pure_virtual'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:vtable for functions::reduce::ReduceFunction<double>: error: undefined reference to '__cxa_pure_virtual'
CMakeFiles/nd4j.dir/cpu/NativeOps.cpp.o:NativeOps.cpp:vtable for functions::reduce::ReduceFunction<double>: error: undefined reference to '__cxa_pure_virtual'
```

What do I need to change to get the missing stuff?

**Comment by andreas-eberle** [comment]
Hi again,

now I get some SIGSEGV. How can I best find out what is causing this?

```
05-30 16:09:25.438 10220-10313/andreaseberle.myapplication A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0xf5d1e0e0 in tid 10313 (Thread-5008)
05-30 16:09:25.438 10220-10312/andreaseberle.myapplication A/libc: Fatal signal 11 (SIGSEGV), code 2, fault addr 0x4630f070 in tid 10312 (Thread-5008)
05-30 16:09:25.438 10220-10314/andreaseberle.myapplication A/libc: Fatal signal 11 (SIGSEGV), code 2, fault addr 0xa572d150 in tid 10314 (Thread-5008)
05-30 16:09:25.478 10220-10289/andreaseberle.myapplication A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x97f00000 in tid 10289 (Thread-5008)
```

**Comment by saudet** [comment]
It sounds like you're not linking with the required libraries. If it were me, I'd just try to use an `android-arm.cmake` file that I know works, such as the one in here:
https://github.com/bytedeco/javacpp-presets/blob/master/artoolkitplus/ARToolKitPlus-2.3.1.patch

But if that doesn't work, then how your building libnd4j?

**Comment by saudet** [comment]
Before we can build libnd4j, we'll need to get OpenBLAS built. That's been a blocker in getting Caffe working on Android as well: https://github.com/BVLC/caffe/pull/2619#issuecomment-164639227

**Comment by BenedictP** [comment]
We've got it working on Android (full credits to @andreas-eberle ):

1. Build OpenBlas inside the Project: https://github.com/sh1r0/caffe-android-lib with this setting: https://github.com/sh1r0/caffe-android-lib#openblas

2. Libnd4j from this repo: https://github.com/andreas-eberle/libnd4j/tree/compile_for_android (Attention: hardcoded paths!).  
./buildnativeoperations.sh (will fail but thats fine)  
cd /blasbuild/cpu/blas  
linking with: https://gist.github.com/BenedictP/0408657f1d66f48a964b5c533ee854f5 (Attention: Fix hardcoded paths!)

3. Checkout https://github.com/bytedeco/javacpp and build it

4. Checkout nd4j from this repo: https://github.com/andreas-eberle/nd4j/tree/android_build (Attention: hardcoded paths!).  
Set LIBND4J_HOME as described in their documentation.  
mvn clean install -DskipTests
will be failing at cuda but thats fine native is already build.  
The needed libs can be found here: "nd4j-backends\nd4j-backend-impls\nd4j-native\target\classes\org\nd4j\nativeblas\android-arm"

5. The *.so object we generated (are in the folder "src/main/jniLibs/armeabi-v7a/"): 
[so_objects.zip](https://github.com/deeplearning4j/libnd4j/files/303032/so_objects.zip)


**But** our problem is now: We're getting different outputs for the same input (unit vector) on an android device vs. a desktop machine.

**Comment by saudet** [comment]
OpenBLAS doesn't officially support `-mfloat-abi=softfp` required by Android's JNI, so I wonder if that hack is actually working properly...

**Comment by andreas-eberle** [comment]
@saudet: Is it possible to use the Eigen BLAS with libnd4j? The caffe project seems to run with it and it seems that Eigen can be compiled for Android (see https://github.com/sh1r0/caffe-android-lib/blob/master/README.md#openblas).

**Comment by saudet** [comment]
@andreas-eberle Anything that implements the BLAS API will work yes.

**Comment by andreas-eberle** [comment]
I was able to get it running with the reference cBLAS implementation. Although that isn't super fast, it works. I now need to fix and document that build process before I will try to compile more advanced BLAS implementations...

**Comment by raver119** [comment]
Please, keep us informed, nd4j on Android is still on our roadmap

**Comment by andreas-eberle** [comment]
I had a look at some other/more advanced BLAS implementations but had some problems:

* OpenBLAS can only be compiled with hard float ABI, which is no longer supported on Android. They think about an Android fix, but it would mean a lot of work (rewriting the kernels).
* clBLAS is not really a BLAS implementation, as it requires manual GPU memory management and so on. So you cannot directly use it in libnd4j.
* The other implementations I saw don't work on Android at all

As the reference BLAS implementation is really slow, I cannot work with it. Therefore I'm currently out of options. Probably the only way to get a decent speedup on Android would be if libnd4j would support OpenCL. For most newer top-level smartphones, OpenCL is supported via external drivers provided by the GPU manufacturers. So although OpenCL is not directly part of Android, you can use it on the phones supporting it. 

Is there a chance that OpenCL will be supported in the future? Or do you have other ideas how to get a big speedup on Android?

**Comment by saudet** [comment]
Actually Android seems to work fine with hard float for me. I've recently created presets for OpenBLAS to provide bindings for the whole of LAPACK that we plan to use in ND4J along with CBLAS:
https://github.com/bytedeco/javacpp-presets/tree/master/openblas
It would be great if you could check it out and let me know of any issues you encounter with that!

**Comment by andreas-eberle** [comment]
Which ndk version did you use? According to several tickets/discussion in the OpenBLAS project it uses hard float ABI for their kernels. However, with NDK 12, you cannot compile for hard float abi anymore.

**Comment by saudet** [comment]
Right, it works with r11c. They are also dropping support for GCC. That's another issue. They never provided the Fortran compiler either. Eventually I guess we'll need to maintain our own toolchain :)

**Comment by raver119** [comment]
I wonder, maybe pure java backend with limited blas impl (dot, axpy, gemm, gemv) support would be viable solution for android and other IoT things?

**Comment by andreas-eberle** [comment]
@saudet: Well, hard float abi has been completely removed. The fortran compiler can be added with a simple flag... Regardless of that, I cannot rely on an outdated build platform :-/ 

@raver119: I have no idea how fast that would be and if that would be enough. However, as the reference BLAS implementation (which is written in C/Fortran is too slow, I don't think a simple java implementation would be fast enough.

**Comment by raver119** [comment]
Another idea here.

When i was adding half support for CUDA, i've checked what we actually using in DL pipeline. And basically only external call is gemm. So another solution is add gemm to NativeOps for cpu, as fallback if no BLAS is available.

And on top of that, i'm going to add gemm anyway, since i need it for half support on Xeon Phi. All other required ops (Axpy, Dot, Asum) - already available as NativeOps.

**Comment by saudet** [comment]
I've looked into OpenBLAS some more and it turns out they are not using NEON (SIMD) instructions in the case of the ARMV7 target anyway, so I figured we might not lose that much performance just compiling the pure C version of the source code. That's what it does for the ARMV5 target and it builds and runs fine with no hard float in sight: https://github.com/bytedeco/javacpp-presets/commit/4d91135c8401e7744ef84b07d2c494f5880a9247

I've tested this quickly in the Android emulator with `cblas_sgemm()` and `cblas_dgemm()` on small (16x16) and large (1024x1024) matrices, but I'm not noticing any drop in performance. @andreas-eberle What about you?

**Comment by saudet** [comment]
Precompiled binaries are also available for 0.6.0 on Maven:
https://repo1.maven.org/maven2/org/nd4j/nd4j-native/0.6.0/
But ND4J 0.6.0 has a check for 64-bit platforms and refuses to run otherwise.
We can temporarily work around that by setting the appropriate property:
```java
System.setProperty("sun.arch.data.model", "64");
```
I have only lightly tested on Android, seems to work fine, but please do report issues you encounter. Thank you all for your time on this!

**Comment by jounih** [comment]
Check this out. Uses NEON and works on Android

https://github.com/Maratyszcza/NNPACK
